File: oldcontest/meme/solutions/carlo.py
#!/usr/bin/env python3

"""
@author: Carlo Malagnino

Task: meme
"""

T=int(input())

for t in range(1,T+1):
    ln=input()
    ln=input().split()
    N=int(ln[0])
    m=int(ln[1])
    
    w=[0]
    p=[0]
    
    for i in range(N):
        ln=input().split()
        w.append(int(ln[0]))
        p.append(int(ln[1]))
    
    M = [ [ (0,-1) for i in range(m+1) ] for j in range(N+1) ]
    
    for i in range(1,N+1):
        for j in range(m+1):
            if(w[i]>j):
                M[i][j]=M[i-1][j]
            else:
                if(p[i]+M[i-1][j-w[i]][0]>M[i-1][j][0]):
                    M[i][j]=(p[i]+M[i-1][j-w[i]][0], i)
                else:
                    M[i][j]=M[i-1][j]
    g=N
    h=m
    ris=[]
    '''
    for i in range(0,N+1):
        for j in range(m+1):
            print(M[i][j], end=" ")
        print()
    '''
    while(M[g][h][1]!=-1):
        meme=M[g][h][1]
        ris.append(meme)
        h-=w[meme]
        g=meme-1
    ris.reverse()
    print("Case #{}: ".format(t), end="")
    for e in ris:
        print(e, end=" ")
    print()

File: contest21old/meme/solutions/carlo.cpp
//Author: Carlo Malagnino

//Task: meme

#include<bits/stdc++.h>

using namespace std;

int main(){
    int T;

    scanf("%d", &T);

    for(int t=1;t<=T;t++){
        int N, m;
        scanf("%d%d", &N, &m);
        int w[N+1], p[N+1];
        for(int i=1;i<=N;i++){
            scanf("%d%d", &w[i], &p[i]);
        }
        pair<int,int> M[N+1][m+1];
        for(int i=0;i<=m;i++){
            M[0][i]={0,-1};
        }
        for(int i=1;i<=N;i++){
            for(int j=0;j<=m;j++){
                if(w[i]>j){
                    M[i][j]=M[i-1][j];
                }
                else{
                    if(p[i]+M[i-1][j-w[i]].first>M[i-1][j].first){
                        M[i][j]={p[i]+M[i-1][j-w[i]].first, i};
                    }
                    else{
                        M[i][j]=M[i-1][j];
                    }
                }
            }
        }
        int g=N, h=m, meme;
        stack<int> ris;
        while(M[g][h].second!=-1){
            meme=M[g][h].second;
            ris.push(meme);
            h-=w[meme];
            g=meme-1;
        }
        printf("Case #%d:", t);
        while(!ris.empty()){
            printf(" %d", ris.top());
            ris.pop();
        }
        printf("\n");
    }
}
